### PR TITLE
Use std::string_view as RowContainer's reference structure for complex types.

### PR DESCRIPTION
**Pull request by usurai**
This PR replaces `StringView` with `std::string_view` as `RowContainer`'s data structure to represent complex type data.

`RowContainer` uses `StringView` as the data structure to represent the reference to variable width data types: `VARCHAR` and complex types like `ARRAY`, `MAP` AND `ROW`. While it makes sense for `VARCHAR`, it does not fit so well with complex types:
- It might lead to confusion as `StringView` suggest a value of type VARCHAR, while for complex types, all they need to store is a pointer to data and size.
- The benefits of `StringView` do not apply to complex types:
    - `StringView` has 4-byte prefix for fast failing of comparison, which is not used for complex types.
    - `StringView` stores data inline for data with size <= 12 , but complex-typed data is bigger than this.
- Construction of `StringView` involves copying first 4 bytes to prefix, which is a waste for complex types.

### CODE REVIEW

File: velox/exec/RowContainer.cpp [hunk 1]
```diff
@@ -135,20 +135,20 @@ RowContainer::RowContainer(
...
-  // is followed by a free bit which is set if the row is in a free
-  // list. The accumulators come next, with size given by
-  // Aggregate::accumulatorFixedWidthSize(). Dependent fields follow.
-  // These are non-key columns for hash join or order by. If there are variable
-  // length columns or accumulators, i.e. ones that allocate extra space, this
-  // space is tracked by a uint32_t after the dependent columns. If this is a
-  // hash join build side, the pointer to the next row with the same key is
-  // after the optional row size.
+  // StringView(for VARCHAR) and std::string_view(ARRAY, MAP and ROW) that
+  // inlines or points to the data.  The number of bytes used by each key is
```
Comment (mbasmanova) [inline velox/exec/RowContainer.cpp:139]: > that inlines or points to the data

This applies only to VARCHAR, not to complex types, no?

**Review by mbasmanova — APPROVED** [review]
@usurai Thank you for the refactoring. This seems clearer to me.

CC: @xiaoxmeng @Yuhta

File: velox/exec/RowContainer.cpp (hunk 1 above)
Comment (usurai) [reply velox/exec/RowContainer.cpp]: Yes, std::string_view won't do inlining.
I've fixed, thanks!

**Review by xiaoxmeng — APPROVED** [review]
@usurai thanks for the change!

**Comment by mbasmanova** [comment]
@usurai Turns out this change is incorrect. Neither StringView not std::string_view can be used as fixed-width complex value in the 'row'. Both StringView and std::string_view imply that the buffer pointed to is a contiguous buffer, but in this case the buffer may or may not be contiguous. 

Also, the implementation of typeKindSize assumes that complex types use StringView and it happens to work with std::string_view only because sizeof(std::string_view) happens to be same as sizeof(StringView), i.e. 16. 

To avoid confusion and prevent bugs caused by it, we may want to introduce a different type to represent a size + pointer to possibly multi-part buffer backed by HashStringAllocator.

CC: @oerling @xiaoxmeng @Yuhta

**Comment by usurai** [comment]
@mbasmanova My understand was: `StringView`/`string_view` in the context of `RowContainer` only act as a reference to the first block of the potential multi-part buffer. With them pointing to the first block, it's `RowContainer` / `HashStringAllocator`'s responsibility to handle non-contiguous parts.

> To avoid confusion and prevent bugs caused by it

Could you suggest potential bugs caused by this? It would be helpful when considering the design of the new type for referencing the potential multi-part buffer. Thanks!

**Comment by usurai** [comment]
@mbasmanova Some update:

> Both StringView and std::string_view imply that the buffer pointed to is a contiguous buffer, but in this case the buffer may or may not be contiguous.

Seems the buffer is always contiguous because:
1. `AllocationPool` which is backing `HashStringAllocator` only does contiguous allocation.
2. If `ByteStream` is backed by `HashStringAllocator`, it only uses the first run of the allocation.

### `AllocationPool` only does contiguous allocation
As we know non-inline varchar/complex type values are stored by `HashStringAllocator` which allocates memory either directly from memory pool(>= 3KB) or from `AllocationPool`. 
In case of `AllocationPool`:
1. If it has already allocated >= 256KB or the allocation request memory size > max size class, calls `allocateContiguous`.
4. Otherwise, calls `allocateNonContiguous`, but specifies `minSizeClass` to equal to `numPages`, that is, asking for a contiguous memory:
https://github.com/facebookincubator/velox/blob/6775fd95ce9188d50269e334409a1d811949d067/velox/common/memory/AllocationPool.cpp#L135-L136

### `HashStringAllocator` backing `ByteStream` only uses the first run of the allocation
`ByteStream::extend` is used when there is no enough space in current `ByteRange` for the appended data. Within `ByteStream::extend`, it calls `StreamArena::newRange` for more memory. Only the base class `StreamArena` uses next run, `HashStringAllocator` directly `allocate` new memory.

Given this, I think it's fine to just use StringView/string_view as the reference type? If so

> the implementation of typeKindSize assumes that complex types use StringView and it happens to work with std::string_view only because sizeof(std::string_view) happens to be same as sizeof(StringView), i.e. 16.

I can fix this by updating `kindToFlatVector` implementation of complex types.

**Comment by mbasmanova** [comment]
@usurai 

> the buffer is always contiguous because

This is not the case. Check out TEST_F(HashStringAllocatorTest, multipart) in velox/common/memory/HashStringAllocator.cpp

**Comment by usurai** [comment]
@mbasmanova Oh I see. I messed up..
Yes, though each call to `allocateNonContiguous` returns allocation with only 1 run, when extend, it gets a new chunk of memory, which makes it non-contiguous.
Thanks for pointing me to the test!


Ok, back to the topic. We want to introduce a new type
- with size + pointer as data
- with StringView's inline functionality
- with a name that indicates it's multi-part characteristic.
Right?

**Pull request closed without merging**